File: 프로그래머스/4/43236. 징검다리/징검다리.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

vector<int> dist;

int solution(int distance, vector<int> rocks, int n) {
    int answer = 0;
    
    sort(rocks.begin(), rocks.end());
    
    dist.push_back(rocks[0]);
    for(int i = 0; i < rocks.size(); ++i){
        if(i != rocks.size() - 1) dist.push_back(rocks[i+1] - rocks[i]);
        else dist.push_back(distance - rocks[i]);
    }
    
    // 굳이 left 자세히 찾을 필요 없는 것 같은..
    int left = 0, right = distance;
    
    while(left < right){
        // 거리의 최솟값
        int mid = (left + right) / 2;
        
        int tmp = 0, cnt = 0;
        for(int i = 0; i < dist.size(); ++i){
            tmp += dist[i];
            
            if(tmp < mid){
                cnt++;
            }else{
                tmp = 0;
            }
        }
        
        // 불가능 : 현재 거리로 최대값을 하려면 바위를 더 제거해야함
        if(cnt > n){
            right = mid;
        }
        // 가능
        else {
            left = mid + 1;
        }
    }
    
    // 모든 바위 제거하는 경우
    if(left == distance) answer = distance;
    else answer = left - 1;
    
    return answer;
}

File: 프로그래머스/4/43236. 징검다리/징검다리.java
import java.util.*;

class Solution {
    
    int[] stone;
    
    public int solution(int distance, int[] rocks, int n) {
        int answer = 0;

        stone = new int[rocks.length + 2];
        for(int i = 0; i < rocks.length; i++) {
            stone[i] = rocks[i];
        }
        stone[rocks.length] = 0;
        stone[rocks.length + 1] = distance;
        
        Arrays.sort(stone);
        
        int left = 1;
        int right = distance + 1;
        
        while(left < right) {
            int mid = (left + right) / 2;
            
            if(simul(mid, n)) {
                left = mid + 1;
                answer = mid;
            }
            else {
                right = mid;
            }
        }
        
        return answer;
    }
    
    // 최소값 k 가능? / n개까지 제거 가능
    boolean simul(int k, int n) {
        int cur = 0;
        for(int i = 1; i < stone.length; i++) {
            cur += stone[i] - stone[i - 1];
            
            // 바위 제거
            if(cur < k) {
                if(--n < 0) return false;
            }
            else {
                cur = 0;
            }
        }
        
        return true;
    }
}